### PR TITLE
Add a function to list linked artifact information

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

Currently, users do not have the direct access to the artifact information from study or trial, 
In this PR, I added a user API to list the artifact information connected to the provided trial or study. 

## Description of the changes

- Add a function to list the information
- Add unittests for this functionality.

### CODE REVIEW

**Comment by not522** [comment]
@c-bata @eukaryo Could you review this PR?

**Comment by c-bata** [comment]
@nabenabe0928 Let me mark this PR as a draft since it's not yet ready for review. Could you please ping me once you have updated?

**Comment by nabenabe0928** [comment]
@c-bata The PR is ready for review!

**Comment by nabenabe0928** [comment]
The conflict has been resolved.

@eukaryo @c-bata 
Could you review this PR?

**Review by eukaryo — APPROVED** [review]
LGTM, mainly about quality of code.
My understanding is that updating [the tutorial](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/012_artifact_tutorial.html) is outside the scope of this PR.

**Review by c-bata — COMMENTED** [review]
Thank you for your pull request! Let me leave some comments.

@porink0424 Could you please review this PR? I will merge this PR after you approved.

**Review by porink0424 — APPROVED** [review]
LGTM👍

@c-bata Please merge this PR after resolving whether `ArtifactMeta` should be made to public or not.

File: optuna/artifacts/__init__.py [hunk 1]
```diff
@@ -10,5 +11,6 @@
     "Boto3ArtifactStore",
     "GCSArtifactStore",
     "Backoff",
+    "get_all_artifact_meta",
```
[the remark this answers is not in the document]
Comment (nabenabe0928) [reply optuna/artifacts/__init__.py:15]: Point taken!
I will add it here!

File: optuna/artifacts/__init__.py [hunk 2]
```diff
@@ -3,14 +3,18 @@
 from optuna.artifacts._download import download_artifact
 from optuna.artifacts._filesystem import FileSystemArtifactStore
 from optuna.artifacts._gcs import GCSArtifactStore
+from optuna.artifacts._list_artifact_meta import get_all_artifact_meta
+from optuna.artifacts._upload import ArtifactMeta
 from optuna.artifacts._upload import upload_artifact
 
 
 __all__ = [
+    "ArtifactMeta",
```
Comment (c-bata) [inline optuna/artifacts/__init__.py:12]: How about listing `ArtifactMeta` in the document as well?

File: tests/artifacts_tests/test_list_artifact_meta.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        artifact_store,
+        storage=storage,
+        mimetype=mimetype,
+        encoding=encoding,
+    )
+    mimetype = "text/plain" if mimetype is None else mimetype
+    if isinstance(study_or_trial, FrozenTrial):
+        artifact_meta_list = get_all_artifact_meta(study_or_trial, storage=storage)
+    else:
+        artifact_meta_list = get_all_artifact_meta(study_or_trial)
```
Comment (c-bata) [inline tests/artifacts_tests/test_list_artifact_meta.py:47]: It looks like a kind of conditional test logic which is described at xUnit Test Patterns.
http://xunitpatterns.com/Conditional%20Test%20Logic.html

Is it possible to simplify this logic?

Comment (nabenabe0928) [reply tests/artifacts_tests/test_list_artifact_meta.py]: Thank you for the suggestion!

Done!

**Pull request merged**